pkg/resource/interfaces.go

```diff
@@ -139,20 +144,39 @@ type Managed interface {
 // A Provider is a Kubernetes object that refers to credentials to connect
 // to an external system.
 type Provider interface {
-	runtime.Object
-	metav1.Object
+	Object
 
 	CredentialsSecretReferencer
 }
 
 // A Target is a Kubernetes object that refers to credentials to connect
 // to a deployment target. Target is a subset of the Claim interface.
 type Target interface {
-	runtime.Object
-	metav1.Object
+	Object
 
 	LocalConnectionSecretWriterTo
 	ManagedResourceReferencer
 
 	Conditioned
 }
+
+// A WorkloadReferencer is a type that has a Workload reference.
+type WorkloadReferencer interface {
```

Comment:
This is what is causing our circular dependency. The immediate options I can think of for solving this problem are:

- Create a separate `crossplane-oam-runtime` repo for the reconcilers and interfaces added in this PR that is not imported by `crossplane`
- Move the OAM types out of core `crossplane` and let both `crossplane` and `crossplane-runtime` import them

Reply:
https://github.com/crossplane/crossplane/blob/206995/apis/oam/v1alpha2/core_types.go#L227

WorkloadReference, TraitReference, and ScopeReference are currently all identical, apart from their Godoc strings. It was pointed out during the original review that they could be a single type.

https://github.com/crossplane/crossplane-runtime/blob/c09220/apis/core/v1alpha1/resource.go#L60

We already have a few subtypes of ObjectReference in crossplane-runtime. I suggest we combine our OAM reference types into one and move it into crossplane-runtime. There's plenty of non-OAM places in the Crossplane codebase where we could use a reference type that requires a GVK and name and has an optional UID. `TypedReference` or something?

Reply:
Great idea! Will update :)